_posts/2020/2020-08-11-why-our-spring-boot-upgrade-was-not-allowed.md

```diff
@@ -0,0 +1,134 @@
+To be honest, we use many of such applications to achieve our goal, but one of such examples is [Spring Boot](https://spring.io/projects/spring-boot){:target="_blank"}, which "makes it easy to create production-grade applications that you can *just run*".
+This does bring some side effects, however.
+
+## Bump!
+
+Google Chrome is built on top of open-source Chromium, which still gets new releases and patches.
+In order to have all the recent security patches and bug fixes, the development team of Google Chrome is forced to upgrade their version of Chromium every once in a while.
+
+Our Beyond framework is built on top of Spring Boot, which still gets new releases and patches.
+There might be some kind of pattern here.
+
+Every once in a while, especially if we have enough capacity to do so, we upgrade our dependencies as well.
+Sometimes, the transition to the newer software version might be seamless or, at least, simple enough.
+The other times, however...
+
+## Method not allowed
+
+It was time for one of such upgrades and after successfully upgrading a couple dozen of other microservices to Spring Boot 2.1 (with only a few hiccups), there were no signs of trouble.
+There were only a few microservices left, and the next on the list was the service, which is responsible for storing and accessing the user uploaded images on our platform.
+
+Every single microservice before that was upgraded with just a minor change to `application.yml`:
+```yaml
+spring:
+  main:
+    allow-bean-definition-overriding: true
+```
+You add that change, modify the microservice to use the needed version of Spring Boot and voilà!
+
+After the same procedure with this service, it successfully started.
+But it was not the same anymore.
+
+Every single test in that microservice failed.
+And it failed with a very strange reason:
+```json
+{
+  "status": 405,
+  "error": "Method Not Allowed",
+  "message": "Request method 'GET' not supported",
+  "path": "/api/images"
+}
+```
+It's like there was never an endpoint to get images in an *image storage* microservice.
+By looking deeper into the code, I found out that what the microservice actually does is connecting to another server and storing or fetching the images from it.
+A proxy, if you will.
+And that other server behind the proxy needed [forwarded headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded){:target="_blank"} to function properly.
+Our proxy is built on top of [Zuul](https://github.com/Netflix/zuul){:target="_blank"}, which was sending the needed headers correctly.
+Emphasis on *was*.
+
+## Why does it have to be beans?
+
+The problem is simple.
+We need forwarded headers.
+We also need the Spring Boot version upgrade.
+What do we do?
+
+It's Bean time!
+```java
+@Bean
+public Filter forwardedHeaderFilter() {
+    return new ForwardedHeaderFilter();
+}
+```
+
+With this bean in configuration class, I ran the tests again, and they all successfully passed.
```

Comment:
Do you have a link we could add here for more information on bean?

Reply:
The general information on what Beans are? Or what this specific Bean is?

https://docs.spring.io/spring/docs/4.3.14.RELEASE_to_4.3.15.RELEASE/Spring%20Framework%204.3.15.RELEASE/org/springframework/web/filter/ForwardedHeaderFilter.html

Reply:
Ah okay I thought that these might also be a kind of class or method that have their own documentation. But if there's only the general one, I'm fine with not linking it :)